[MRG+2] make_circles() now works with odd number of samples, test added

christianbraune79 opened the pull request:
#### Reference Issues/PRs
Fixes #10037 and adds corresponding tests

#### What does this implement/fix? Explain your changes.
Fixes he faulty behaviour of ```make_circles``` when ```n_samples``` is an odd number.
Adds test ```test_make_circles``` to ```datasets/tests/test_samples_generator.py```.

#### Any other comments?

cbrauneovgude committed:
fixes #10037

cbrauneovgude committed:
tests for #10037, tests for odd number of samples and whether generated points lie on the expected circles similar to test_make_moons()

cbrauneovgude committed:
Merge branch 'make_circles' of https://github.com/christianbraune79/scikit-learn into make_circles

cbrauneovgude committed:
nasty doubled lines of code removed

TomDLT commented on sklearn/datasets/samples_generator.py:
```diff
@@ -613,20 +613,23 @@ def make_circles(n_samples=100, shuffle=True, noise=None, random_state=None,
-    inner_circ_y = outer_circ_y * factor
+    # so as not to have the first point = last point, we set endpoint=False
+    linspace_out = np.linspace(0, 2 * np.pi, n_samples_out, endpoint=False)
+    linspace_in = np.linspace(0, 2 * np.pi, n_samples_in, endpoint=True)
```
Why `endpoint=True` in the second?

christianbraune79 replied:
Thanks. I feel like I thought it might have been necessary, but I cannot think of a proper reason right now. I'll chnage it to ```endpoint=False```. Thanks.

TomDLT reviewed (approved):
LGTM

rth commented on sklearn/datasets/tests/test_samples_generator.py:
```diff
@@ -385,3 +386,16 @@ def test_make_moons():
+
+def test_make_circles():
+    f = 0.3
+    X, y = make_circles(7, shuffle=False, noise=None, factor=f)
```
Thank you for the PR! 

It might be useful to test with both `n_samples` odd and even, to make sure this change doesn't break the previous behaviour.

christianbraune79 replied:
Thanks for the advice!
I'll also add an additional test then, which tests whether the samples were correctly distributed across both classes.

jnothman replied:
Apart from not testing the even case, I think the current test looks fine...?

qinhanmin2014 commented on sklearn/datasets/samples_generator.py:
```diff
@@ -613,20 +613,23 @@ def make_circles(n_samples=100, shuffle=True, noise=None, random_state=None,
 
     if factor > 1 or factor < 0:
         raise ValueError("'factor' has to be between 0 and 1.")
+        
```
./sklearn/datasets/samples_generator.py:616:1: W293 blank line contains whitespace

christianbraune79 replied:
Thanks. 
It seems PyCharm is not the solution to every PEP8-related issue. :/

qinhanmin2014 reviewed (commented):
Please fix the PEP8 error to pass the test :)

cbrauneovgude committed:
changes according to comments in PR

cbrauneovgude committed:
assert_equal obviously has a different signature.

amueller commented on sklearn/datasets/samples_generator.py:
```diff
@@ -614,19 +614,22 @@ def make_circles(n_samples=100, shuffle=True, noise=None, random_state=None,
     if factor > 1 or factor < 0:
         raise ValueError("'factor' has to be between 0 and 1.")
```
do you want to add a test for this line, too? ;)

christianbraune79 replied:
I actually never understood, why a ```factor > 1``` raises a ```ValueError```. All it would to is turn the inner circle into the outer one, allowing me to generate the samples more intuitively. Compare "the one circle's radius is 0.142 of the other one's" vs. "Bigger circler is 7 times the size of the smaller one".

It would remove the dataset's property to be (roughly, due to noise) within a [-1, 1]² box, though...

jnothman replied:
I think leave it as is. Test if you will

qinhanmin2014 replied:
@christianbraune79 
I think `0 < factor < 1` might be enough. From my perspective, you can use factor to specify how much bigger is the big circle and apply operations on X to specify the position of the big circle (and the small circle). For the example you mentioned, you might use ``X, y = make_circle(factor=1/7); X=7*X`` to achieve your goal. Even if you allow `factor > 1`, you still can't meet the need of many situations, because the center of two circles are fixed to (0,0) and the radius of one circle is fixed to 1.
Also, as the core dev suggested, it will be better if you can add a test to check the error is raised.

amueller commented on sklearn/datasets/samples_generator.py:
```diff
@@ -614,19 +614,22 @@ def make_circles(n_samples=100, shuffle=True, noise=None, random_state=None,
-    outer_circ_y = np.sin(linspace)
-    inner_circ_x = outer_circ_x * factor
-    inner_circ_y = outer_circ_y * factor
+    # so as not to have the first point = last point, we set endpoint=False
```
You're so much smarter than me....

amueller commented on sklearn/datasets/tests/test_samples_generator.py:
```diff
@@ -385,3 +386,23 @@ def test_make_moons():
+    assert_equal(X.shape, (10, 2), "X shape mismatch")
+    assert_equal(y.shape, (10,), "y shape mismatch")
+
+    assert_equal(X[y == 0].shape, (5, 2),
```
If you want you could add a test that it's 3 and 4 for the test above? maybe that it's at most one different from 7 // 2? but fine either way.

amueller reviewed (approved):
you could add another test, but looks good either way.

jnothman commented:
Please add an entry to the change log at `doc/whats_new`.

cbrauneovgude committed:
Adjusted documentation for make_circles
Added a test to check if really only factors in (0, 1) (excluding borders) are accepted
Adjusted factor check (1.0 was accepted before, though doc said otherwise)

cbrauneovgude committed:
added entry under "Decomposition, ..." as another datasets-related bug was already mentioned in there.

qinhanmin2014 commented on sklearn/datasets/samples_generator.py:
```diff
@@ -611,22 +612,25 @@ def make_circles(n_samples=100, shuffle=True, noise=None, random_state=None,
     """
 
-    if factor > 1 or factor < 0:
+    if factor >= 1 or factor < 0:
```
Personally, I'm with you. But this latest source code change is not reviewed by anyone else. So need to confirm with core devs to see whether we need factor=1 in extreme case.

christianbraune79 replied:
Yes, definitely. If it was a typo in the docs then I'll revert the change and adjust the documentation. :)

jnothman replied:
I this is fine.

qinhanmin2014 reviewed (approved):
LGTM except for something to confirm from core devs. Also wondering if it will be helpful to explicitly state that the center of the two circles is fixed to (0,0) and the radius of the outer circle is 1.

jnothman commented on doc/whats_new/_contributors.rst:
```diff
@@ -141,3 +141,5 @@
 
 .. _Arthur Mensch: https://amensch.fr
+
+.. _Christian Braune: https://github.com/christianbraune79
```
This is unused and unnecessary. These says we try to only use this format for core devs...

jnothman reviewed (approved):
Otherwise LGTM

qinhanmin2014 commented:
@christianbraune79 Could you please take some time to finish this, it's very close form merge.
(1) please remove your name in _contributors.rst, that's for core devs
(2) please extend the test to even case, you might consider a loop to avoid duplicate code. Also consider to add a comment to explain that we are testing both odd case and even case here.

cbrauneovgude committed:
removed wrong entry

cbrauneovgude committed:
all tests for odd and even case

cbrauneovgude committed:
added final comment

christianbraune79 commented:
@qinhanmin2014 
Done. Sorry for (1) taking so long to respond (family stuff) and (2) putting myself in the contributors list. At least now I learned, what it is for. :)

cbrauneovgude committed:
'107 > 79' fixed

glemaitre commented on sklearn/datasets/tests/test_samples_generator.py:
```diff
@@ -385,3 +386,28 @@ def test_make_moons():
+
+
+def test_make_circles():
+    f = 0.3
```
can you call this `factor` instead of `f`

glemaitre replied:
or you can even affect 0.3 to factor inside the function itself.

glemaitre commented on sklearn/datasets/tests/test_samples_generator.py:
```diff
@@ -385,3 +386,28 @@ def test_make_moons():
+def test_make_circles():
+    f = 0.3
+
+    # Testing odd and even case, because in the past make_circles always
```
you can put this comment just below the def

glemaitre reviewed (commented):
3 nitpicks more :)

cbrauneovgude committed:
refactoring f into factor

cbrauneovgude committed:
pep8

qinhanmin2014 commented:
@jnothman @amueller Could you please give a final check? I think it is OK for merge. Thanks :)

qinhanmin2014 commented:
@christianbraune79 Thanks for the issue and the PR :)
